gateway/ln-gateway/src/ng/register.rs

```diff
@@ -0,0 +1,260 @@
+use fedimint_ln_common::LightningGateway;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tracing::error;
+use url::Url;
+
+use super::{GatewayClientContext, INITIAL_REGISTER_BACKOFF_DURATION};
+use crate::db::FederationRegistrationKey;
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// State machine that registers the lightning gateway with
+/// the federation.
+///
+/// ```mermaid
+/// graph LR
+/// classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///    Register -- register lightning gateway succeeded --> WaitForTTL
+///    Register -- register lightning gateway failed --> FailureBackoff
+///    FailureBackoff -- wait for backoff duration --> Register
+///    WaitForTTL -- wait for time to live to expire and `LightningGateway` has not changed --> Register
+///    WaitForTTL -- wait for time to live to expire and `LightningGateway` has changed --> Done
+/// ```
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub enum RegisterFederationStates {
+    Register(RegisterGateway),
+    WaitForTTL(WaitForTimeToLive),
+    FailureBackoff(FailureBackoff),
+    Done,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct RegisterFederationCommon {
+    pub operation_id: OperationId,
+    pub time_to_live: Duration,
+    pub registration_info: LightningGateway,
+    pub federation_id: FederationId,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct RegisterFederationStateMachine {
+    pub common: RegisterFederationCommon,
+    pub state: RegisterFederationStates,
+}
+
+impl State for RegisterFederationStateMachine {
+    type ModuleContext = GatewayClientContext;
+    type GlobalContext = DynGlobalClientContext;
+
+    fn transitions(
+        &self,
+        _context: &Self::ModuleContext,
+        global_context: &Self::GlobalContext,
+    ) -> Vec<fedimint_client::sm::StateTransition<Self>> {
+        match &self.state {
+            RegisterFederationStates::Register(register_gateway) => {
+                register_gateway.transitions(global_context.clone(), self.common.clone())
+            }
+            RegisterFederationStates::WaitForTTL(wait_for_ttl) => {
+                wait_for_ttl.transitions(self.common.clone())
+            }
+            RegisterFederationStates::FailureBackoff(failure_backoff) => {
+                failure_backoff.transitions(self.common.clone())
+            }
+            RegisterFederationStates::Done => {
+                vec![]
+            }
+        }
+    }
+
+    fn operation_id(&self) -> fedimint_client::sm::OperationId {
+        self.common.operation_id
+    }
+}
+
+#[derive(Error, Debug, Serialize, Deserialize, Encodable, Decodable, Clone, Eq, PartialEq)]
+pub enum GatewayRegistrationError {
+    #[error("Error registering the Lightning Gateway. Gateway API: {api:?}")]
+    RegistrationError { api: Url },
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Decodable, Encodable)]
+pub struct RegisterGateway {
```

Comment:
nit: rename to `RegisterFederation`

Reply:
Does `RegisterFederation` make more sense? This operation is actually registering the gateway _with_ the federation. This seems to imply that it's registering the federation somehow.

Reply:
I think I might just change it to `RegisterWithFederation`